gen/lib/gen/generator.rb

```diff
@@ -0,0 +1,113 @@
+require 'gen'
+require 'fileutils'
+require 'pathname'
+
+module Gen
+  class Generator
+    def self.run(project_name)
+      new(project_name).run
+    end
+
+    TEMPLATE_ROOT = File.expand_path('gen/template', Gen::ROOT)
+
+    VALID_PROJECT_NAME = /\A[a-z][a-z0-9]*\z/
+    private_constant :VALID_PROJECT_NAME
+
+    VENDOR_TRANSLATIONS = {
+      'Gemfile'            => false,
+      'exe/__app__-gems'   => false,
+      'exe/__app__-vendor' => 'exe/__app__',
+    }.freeze
+    private_constant :VENDOR_TRANSLATIONS
+
+    BUNDLER_TRANSLATIONS = {
+      'bin'                => false,
+      'bin/update-deps'    => false,
+      'exe/__app__-gems'   => 'exe/__app__',
+      'exe/__app__-vendor' => false,
+    }.freeze
+    private_constant :BUNDLER_TRANSLATIONS
+
+    def initialize(project_name)
+      raise(
+        CLI::Kit::Abort,
+        "project name must match {{bold:#{VALID_PROJECT_NAME}}} (but can be changed later)"
+      ) unless project_name =~ VALID_PROJECT_NAME
+      @project_name = project_name
+      @title_case = @project_name[0].upcase + @project_name[1..-1]
+    end
+
+    def run
+      vendor = ask_vendor?
+      create_project_dir
+      if vendor
+        copy_files(translations: VENDOR_TRANSLATIONS)
+      else
+        copy_files(translations: BUNDLER_TRANSLATIONS)
+      end
+    end
+
+    private
+
+    def ask_vendor?
+      vendor = nil
+      CLI::UI::Frame.open('Configuration') do
+        vendor = CLI::UI.ask(
+          'How would you like the application to consume {{command:cli-kit}} and {{command:cli-ui}}?',
+          options: %w(vendor bundler)
+        )
+      end
+      vendor == 'vendor'
+    end
+
+    def create_project_dir
+      info(create: '')
+      FileUtils.mkdir(@project_name)
+    rescue Errno::EEXIST
+      error("directory already exists: #{@project_name}")
+    end
+
+    def copy_files(translations:)
+      each_template_file do |source_name|
+        target_name = translations.fetch(source_name, source_name)
+        next if target_name == false
+        target_name = apply_template_variables(target_name)
+
+        source = File.join(TEMPLATE_ROOT, source_name)
+        target = File.join(@project_name, target_name)
+
+        info(create: target_name)
+
+        if Dir.exist?(source)
```

Comment:
This confuses me, what are we checking here?

Reply:
Whether it's a directory.

Reply:
... I get that 😆 Why do we need to check it?

Reply:
Because `each_template_file do |source_name|` goes through everything in the templates directory: files and directories.

Reply:
So then if it goes through all things it finds in the directory, then the thing is guaranteed to exist... why do we need to check

Reply:
It's not really an existence check, it's a "is it a directory" check.

Reply:
Then we can use `File.directory?`. IMO it's more clear of the intention